src/message.ts

```diff
@@ -8,145 +6,119 @@ import {
   InsufficientCoinError,
   ValueError,
 } from './error'
+import { MsgRequestData as MsgRequestDataProto } from '../proto/oracle/v1/tx_pb'
+import { MsgSend as MsgSendProto } from '../proto/cosmos/bank/v1beta1/tx_pb'
+import { MsgDelegate as MsgDelegateProto } from '../proto/cosmos/staking/v1beta1/tx_pb'
+import { Coin } from '../proto/cosmos/base/v1beta1/coin_pb'
+import { Any } from 'google-protobuf/google/protobuf/any_pb'
 
-export abstract class Msg {
-  abstract asJson(): { type: string; value: any }
-  abstract getSender(): Address
-  abstract validate(): boolean
-}
-export class MsgRequest extends Msg {
-  oracleScriptID: number
-  calldata: Buffer
-  askCount: number
-  minCount: number
-  clientID: string
-  sender: Address
-
+export class MsgRequestData extends MsgRequestDataProto {
   constructor(
-    oracleScriptID: number,
+    oracleScriptId: number,
     calldata: Buffer,
     askCount: number,
     minCount: number,
-    clientID: string,
-    sender: Address,
+    clientId: string,
+    sender: string,
+    fee: Coin[] = [],
+    prepareGas: number = 50000,
+    executeGas: number = 300000,  
   ) {
     super()
-
-    this.oracleScriptID = oracleScriptID
-    this.calldata = calldata
-    this.askCount = askCount
-    this.minCount = minCount
-    this.clientID = clientID
-    this.sender = sender
+    this.setOracleScriptId(oracleScriptId)
+    this.setCalldata(calldata)
+    this.setAskCount(askCount)
+    this.setMinCount(minCount)
+    this.setClientId(clientId)
+    this.setFeeLimitList(fee)
+    this.setPrepareGas(prepareGas)
+    this.setExecuteGas(executeGas)
+    this.setSender(sender)
   }
 
-  asJson() {
-    return {
-      type: 'oracle/Request',
-      value: {
-        ask_count: this.askCount.toString(),
-        calldata: this.calldata.toString('base64'),
-        client_id: this.clientID,
-        min_count: this.minCount.toString(),
-        oracle_script_id: String(this.oracleScriptID),
-        sender: this.sender.toAccBech32(),
-      },
-    }
-  }
 
-  getSender() {
-    return this.sender
+  toAny() {
+    if (this.validate()) {
+      const anyMsg = new Any()
+      const name = 'oracle.v1.MsgRequestData'
+      anyMsg.pack(this.serializeBinary(), name, '/')
+      return anyMsg  
+    }
+    return undefined
   }
 
   validate() {
```

Comment:
also check on feelimit that should be positive.

Reply:
Since amount in `Fee` is a string, the code might be kinda ugly

Reply:
Should use `coin.getAmoun()` instead of `coin.toArray()` to get the amount, then convert to Number using `Number()`.

Reply:
😬  My bad, forgot that we have that method